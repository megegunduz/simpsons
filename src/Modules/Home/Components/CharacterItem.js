import React from 'react';
import { TouchableOpacity, Text, View, Image } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { useDispatch } from 'react-redux';

import { HomeSlice } from '../Redux/HomeRedux';
import { Icon } from '../../../Components';
import { ScreenNames, Svgs } from '../../../Constants';

import styles from '../styles/CharacterItemStyles';

/**
 *
 * @param {object} props
 * @param {import('../../../Types/Character').Character} props.character
 * @param {number} props.index
 * @returns {React.Component}
 */
const CharacterItem = props => {
    const { character } = props;

    const { avatar, name, ranking } = character;

    const dispatch = useDispatch();
    const navigation = useNavigation();

    const _onPress_Navigate = () => {
        navigation.navigate(ScreenNames.DETAIL, {
            character,
        });
    };

    const _onPress_Delete = () => {
        dispatch(HomeSlice.actions.deleteCharacter({ character }));
    };

    const _onPress_Up = () => {
        dispatch(
            HomeSlice.actions.updateRanking({
                character,
                updateType: 'increase',
            }),
        );
    };

    const _onPress_Down = () => {
        dispatch(
            HomeSlice.actions.updateRanking({
                character,
                updateType: 'decrease',
            }),
        );
    };

    return (
        <View style={styles.container}>
            <TouchableOpacity
                style={styles.touchable}
                onPress={_onPress_Navigate}
            >
                <Text style={styles.ranking}>{ranking}</Text>
                <Image
                    source={{ uri: avatar }}
                    style={styles.image}
                    resizeMode={'contain'}
                />
                <Text style={styles.name} numberOfLines={2}>
                    {name}
                </Text>
                <View style={styles.iconsContainer}>
                    <Icon
                        svg={Svgs.Up}
                        style={styles.upIcon}
                        containerStyle={styles.iconContainer}
                        onPress={_onPress_Up}
                    />
                    <Icon
                        svg={Svgs.Down}
                        style={styles.downIcon}
                        containerStyle={styles.iconContainer}
                        onPress={_onPress_Down}
                    />
                    <Icon
                        svg={Svgs.Trash}
                        style={styles.deleteIcon}
                        containerStyle={styles.iconContainer}
                        onPress={_onPress_Delete}
                    />
                </View>
            </TouchableOpacity>
        </View>
    );
};

export default CharacterItem;
