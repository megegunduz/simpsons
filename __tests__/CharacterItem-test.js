import React from 'react';
import CharacterItem from '../src/Modules/Home//Components/CharacterItem';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import renderer from 'react-test-renderer';

const character = {
    name: 'Homer Simpson',
    avatar: 'https://static.wikia.nocookie.net/simpsons/images/b/bd/Homer_Simpson.png/revision/latest/scale-to-width-down/300?cb=20201222215437',
    job: 'Nuclear Safety Inspector',
    description:
        "Homer Jay Simpson (born May 12, 1956) is the main protagonist and one of the five main characters of The Simpsons series (or show). He is the spouse of Marge Simpson and father of Bart, Lisa and Maggie Simpson. Homer is overweight (said to be ~240 pounds), lazy, and often ignorant to the world around him. Although Homer has many flaws, he has shown to have great caring, love, and even bravery to those he cares about and, sometimes, even others he doesn't. He also serves as the main protagonist of the The Simpsons Movie. He is 39 years old and was born in 1956.",
    id: '14',
};

describe('CharacterItem', () => {
    const initialState = {};
    const mockStore = configureStore();
    let store;

    it('Renders corrently', () => {
        store = mockStore(initialState);
        const instance = renderer.create(
            <Provider store={store}>
                <CharacterItem character={character} index={0} />
            </Provider>,
        );

        expect(instance.toJSON()).toMatchSnapshot();
    });
});
