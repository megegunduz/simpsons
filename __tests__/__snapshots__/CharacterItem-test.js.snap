// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterItem Renders corrently 1`] = `
<View
  style={{}}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "marginRight": 31.250000000000004,
        }
      }
    />
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "https://static.wikia.nocookie.net/simpsons/images/b/bd/Homer_Simpson.png/revision/latest/scale-to-width-down/300?cb=20201222215437",
        }
      }
      style={
        {
          "aspectRatio": 0.5,
          "marginRight": 31.250000000000004,
          "width": 104.16666666666667,
        }
      }
    />
    <Text
      numberOfLines={2}
      style={
        {
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      Homer Simpson
    </Text>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "padding": 10.416666666666668,
          }
        }
        testID="icon-touchable"
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={62.50000000000001}
          bbWidth={62.50000000000001}
          focusable={false}
          height={62.50000000000001}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 62,
                "width": 62,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={62.50000000000001}
          xmlns="http://www.w3.org/2000/svg"
        >
          <RNSVGGroup>
            <RNSVGPath
              d="M12 16q.425 0 .713-.288Q13 15.425 13 15v-3.2l.9.9q.275.275.7.275.425 0 .7-.275.275-.275.275-.7 0-.425-.275-.7l-2.6-2.6q-.15-.15-.325-.213-.175-.062-.375-.062t-.375.062q-.175.063-.325.213l-2.6 2.6q-.275.275-.275.7 0 .425.275.7.275.275.7.275.425 0 .7-.275l.9-.9V15q0 .425.288.712.287.288.712.288Zm0 6q-2.075 0-3.9-.788-1.825-.787-3.175-2.137-1.35-1.35-2.137-3.175Q2 14.075 2 12t.788-3.9q.787-1.825 2.137-3.175 1.35-1.35 3.175-2.138Q9.925 2 12 2t3.9.787q1.825.788 3.175 2.138 1.35 1.35 2.137 3.175Q22 9.925 22 12t-.788 3.9q-.787 1.825-2.137 3.175-1.35 1.35-3.175 2.137Q14.075 22 12 22Zm0-2q3.325 0 5.663-2.337Q20 15.325 20 12t-2.337-5.663Q15.325 4 12 4T6.338 6.337Q4 8.675 4 12t2.338 5.663Q8.675 20 12 20Zm0-8Z"
              fill={4282807086}
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
      <View
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "padding": 10.416666666666668,
          }
        }
        testID="icon-touchable"
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={62.50000000000001}
          bbWidth={62.50000000000001}
          focusable={false}
          height={62.50000000000001}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 62,
                "width": 62,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={62.50000000000001}
          xmlns="http://www.w3.org/2000/svg"
        >
          <RNSVGGroup>
            <RNSVGPath
              d="M12 15.575q.2 0 .375-.063.175-.062.325-.212l2.6-2.6q.275-.275.275-.7 0-.425-.275-.7-.275-.275-.7-.275-.425 0-.7.275l-.9.9V9q0-.425-.287-.713Q12.425 8 12 8t-.712.287Q11 8.575 11 9v3.2l-.9-.9q-.275-.275-.7-.275-.425 0-.7.275-.275.275-.275.7 0 .425.275.7l2.6 2.6q.15.15.325.212.175.063.375.063ZM12 22q-2.075 0-3.9-.788-1.825-.787-3.175-2.137-1.35-1.35-2.137-3.175Q2 14.075 2 12t.788-3.9q.787-1.825 2.137-3.175 1.35-1.35 3.175-2.138Q9.925 2 12 2t3.9.787q1.825.788 3.175 2.138 1.35 1.35 2.137 3.175Q22 9.925 22 12t-.788 3.9q-.787 1.825-2.137 3.175-1.35 1.35-3.175 2.137Q14.075 22 12 22Zm0-2q3.325 0 5.663-2.337Q20 15.325 20 12t-2.337-5.663Q15.325 4 12 4T6.338 6.337Q4 8.675 4 12t2.338 5.663Q8.675 20 12 20Zm0-8Z"
              fill={4291299347}
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
      <View
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "padding": 10.416666666666668,
          }
        }
        testID="icon-touchable"
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={62.50000000000001}
          bbWidth={62.50000000000001}
          focusable={false}
          height={62.50000000000001}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 62,
                "width": 62,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={62.50000000000001}
          xmlns="http://www.w3.org/2000/svg"
        >
          <RNSVGGroup>
            <RNSVGPath
              d="M7 21q-.825 0-1.412-.587Q5 19.825 5 19V6q-.425 0-.713-.287Q4 5.425 4 5t.287-.713Q4.575 4 5 4h4q0-.425.288-.713Q9.575 3 10 3h4q.425 0 .713.287Q15 3.575 15 4h4q.425 0 .712.287Q20 4.575 20 5t-.288.713Q19.425 6 19 6v13q0 .825-.587 1.413Q17.825 21 17 21ZM7 6v13h10V6Zm2 10q0 .425.288.712Q9.575 17 10 17t.713-.288Q11 16.425 11 16V9q0-.425-.287-.713Q10.425 8 10 8t-.712.287Q9 8.575 9 9Zm4 0q0 .425.288.712.287.288.712.288t.713-.288Q15 16.425 15 16V9q0-.425-.287-.713Q14.425 8 14 8t-.712.287Q13 8.575 13 9ZM7 6v13V6Z"
              fill={4279860928}
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
</View>
`;
